pkg/asset/machines/worker.go

```diff
@@ -172,44 +178,60 @@ func (w *Worker) Generate(dependencies asset.Parents) error {
 		return errors.Wrap(err, "failed to create MachineConfig manifests for worker machines")
 	}
 
-	if len(machineSets) == 0 {
-		return nil
-	}
-	list := &metav1.List{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: "v1",
-			Kind:       "List",
-		},
-		Items: make([]runtime.RawExtension, len(machineSets)),
-	}
-	for i, set := range machineSets {
-		list.Items[i] = runtime.RawExtension{Object: set}
-	}
-	data, err = yaml.Marshal(list)
-	if err != nil {
-		return errors.Wrap(err, "failed to marshal")
-	}
-	w.MachineSetFile = &asset.File{
-		Filename: filepath.Join(directory, workerMachineSetFileName),
-		Data:     data,
+	w.MachineSetFiles = make([]*asset.File, len(machineSets))
+	padFormat := fmt.Sprintf("%%0%dd", len(fmt.Sprintf("%d", len(machineSets))))
+	for i, machineSet := range machineSets {
+		data, err := yaml.Marshal(machineSet)
+		if err != nil {
+			return errors.Wrapf(err, "marshal worker %d", i)
+		}
+
+		padded := fmt.Sprintf(padFormat, i)
+		w.MachineSetFiles[i] = &asset.File{
+			Filename: filepath.Join(directory, fmt.Sprintf(workerMachineSetFileName, padded)),
+			Data:     data,
+		}
 	}
+
 	return nil
 }
 
 // Files returns the files generated by the asset.
 func (w *Worker) Files() []*asset.File {
-	files := make([]*asset.File, 0, 1+len(w.MachineConfigFiles)+1)
+	files := make([]*asset.File, 0, 1+len(w.MachineConfigFiles)+len(w.MachineSetFiles))
 	if w.UserDataFile != nil {
 		files = append(files, w.UserDataFile)
 	}
 	files = append(files, w.MachineConfigFiles...)
-	if w.MachineSetFile != nil {
-		files = append(files, w.MachineSetFile)
-	}
+	files = append(files, w.MachineSetFiles...)
 	return files
 }
 
-// Load returns false since this asset is not written to disk by the installer.
+// Load reads the asset files from disk.
 func (w *Worker) Load(f asset.FileFetcher) (found bool, err error) {
-	return false, nil
+	file, err := f.FetchByName(filepath.Join(directory, workerUserDataFileName))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	w.UserDataFile = file
+
+	w.MachineConfigFiles, err = machineconfig.Load(f, "worker", directory)
+	if err != nil {
+		return true, err
+	}
+
+	fileList, err := f.FetchByPattern(filepath.Join(directory, workerMachineSetFileNamePattern))
+	if err != nil {
+		return true, err
+	}
+
+	if len(fileList) == 0 {
```

Comment:
zero worker are valid configuration. and userData is still output for them https://github.com/openshift/installer/pull/1481/commits/c0bc4dce154e26f35c3006f785e66f2deb68074e#diff-df3202277ffea159153362ec095a8762L104

Reply:
> zero worker are valid configuration. and userData is still output for them...

Do we want to continue with that approach?  It seems like users who want to provide their own MachineSets can provide their own user data.  I see three cases:

a. User configures at least one non-empty machine pool (the default).  They get MachineSet(s) and user data (and with #1150, MachineConfigs).
b. User configures at least one machine pool, but `replicas` is zero in all of them.  They get the same thing as (a), except `replicas` will be zero, so the machine API won't actually create any workers.
c. User explicitly configures `compute` empty...  And I'm not actually sure how they'd do that, since we have [this defaulting logic][1].

So I guess you are interested in folks who want a multi-step install where they explicitly supply worker user data but no MachineSets?  I'm ok dropping this guard to let that through.  Do you want to support "supplies MachineConfigs but no user data or MachineSets" too?  Currently we're treating anything that skips worker user data as "no overrides supplied".

[1]: https://github.com/openshift/installer/blob/fa089644bde9e062fc3edcfac24187fdbe8bca01/pkg/types/defaults/installconfig.go#L55-L62

Reply:
> > zero worker are valid configuration. and userData is still output for them...
> 
> Do we want to continue with that approach?

zero compute is valid configuration we need to support.

> It seems like users who want to provide their own MachineSets can provide their own user data. I see three cases:
> a. User configures at least one non-empty machine pool (the default). They get MachineSet(s) and user data (and with #1150, MachineConfigs).
> b. User configures at least one machine pool, but `replicas` is zero in all of them. They get the same thing as (a), except `replicas` will be zero, so the machine API won't actually create any workers.
> c. User explicitly configures `compute` empty... And I'm not actually sure how they'd do that, since we have [this defaulting logic](https://github.com/openshift/installer/blob/fa089644bde9e062fc3edcfac24187fdbe8bca01/pkg/types/defaults/installconfig.go#L55-L62).
> 
> So I guess you are interested in folks who want a multi-step install where they explicitly supply worker user data but no MachineSets? I'm ok dropping this guard to let that through. Do you want to support "supplies MachineConfigs but no user data or MachineSets" too? Currently we're treating anything that skips worker user data as "no overrides supplied".

This was possible before:
```console
$(pr-1481) ./bin/openshift-install --dir dev create manifests
$(pr-1481) ls dev/**
dev/manifests:
04-openshift-machine-config-operator.yaml  cluster-ingress-02-config.yml  etcd-service.yaml                 kube-system-configmap-etcd-ca-bundle.yaml          kube-system-secret-etcd-client.yaml         openshift-config-configmap-etcd-metrics-serving-ca.yaml
cluster-config.yaml                        cluster-network-01-crd.yml     host-etcd-service-endpoints.yaml  kube-system-configmap-etcd-serving-ca.yaml         kube-system-secret-etcd-signer-client.yaml  openshift-config-secret-etcd-metrics-client.yaml
cluster-dns-02-config.yml                  cluster-network-02-config.yml  host-etcd-service.yaml            kube-system-configmap-root-ca.yaml                 kube-system-secret-etcd-signer.yaml         pull.json
cluster-infrastructure-02-config.yml       cvo-overrides.yaml             kube-cloud-config.yaml            kube-system-secret-etcd-client-ca-deprecated.yaml  machine-config-server-tls-secret.yaml

dev/openshift:
99_binding-discovery.yaml          99_openshift-cluster-api_master-machines-0.yaml  99_openshift-cluster-api_master-user-data-secret.yaml  99_openshift-cluster-api_worker-machineset-2.yaml      99_openshift-machineconfig_worker.yaml
99_cloud-creds-secret.yaml         99_openshift-cluster-api_master-machines-1.yaml  99_openshift-cluster-api_worker-machineset-0.yaml      99_openshift-cluster-api_worker-user-data-secret.yaml  99_role-cloud-creds-secret-reader.yaml
99_kubeadmin-password-secret.yaml  99_openshift-cluster-api_master-machines-2.yaml  99_openshift-cluster-api_worker-machineset-1.yaml      99_openshift-machineconfig_master.yaml
$(pr-1481) rm dev/openshift/99_openshift-cluster-api_worker-machineset-*.yaml
$(pr-1481) ./bin/openshift-install --dir dev create cluster
FATAL failed to fetch Terraform Variables: failed to load asset "Worker Machines": worker machine manifests are required if you also provide openshift/99_openshift-cluster-api_worker-user-data-secret.yaml
```

```
$(master) ./bin/openshift-install --dir dev create manifests
WARNING There are no compute nodes specified. The cluster will not fully initialize without compute nodes.
INFO Consuming "Install Config" from target directory
$(master) ls dev/**
dev/manifests:
04-openshift-machine-config-operator.yaml  cluster-ingress-02-config.yml  etcd-service.yaml                 kube-system-configmap-etcd-ca-bundle.yaml          kube-system-secret-etcd-client.yaml         openshift-config-configmap-etcd-metrics-serving-ca.yaml
cluster-config.yaml                        cluster-network-01-crd.yml     host-etcd-service-endpoints.yaml  kube-system-configmap-etcd-serving-ca.yaml         kube-system-secret-etcd-signer-client.yaml  openshift-config-secret-etcd-metrics-client.yaml
cluster-dns-02-config.yml                  cluster-network-02-config.yml  host-etcd-service.yaml            kube-system-configmap-root-ca.yaml                 kube-system-secret-etcd-signer.yaml         pull.json
cluster-infrastructure-02-config.yml       cvo-overrides.yaml             kube-cloud-config.yaml            kube-system-secret-etcd-client-ca-deprecated.yaml  machine-config-server-tls-secret.yaml

dev/openshift:
99_binding-discovery.yaml   99_kubeadmin-password-secret.yaml                99_openshift-cluster-api_master-machines-1.yaml  99_openshift-cluster-api_master-user-data-secret.yaml  99_openshift-cluster-api_worker-user-data-secret.yaml
99_cloud-creds-secret.yaml  99_openshift-cluster-api_master-machines-0.yaml  99_openshift-cluster-api_master-machines-2.yaml  99_openshift-cluster-api_worker-machineset.yaml        99_role-cloud-creds-secret-reader.yaml
$(master) rm dev/openshift/99_openshift-cluster-api_worker-machineset.yaml
$(master)  ./bin/openshift-install --dir dev create clusterINFO Consuming "Master Machines" from target directory
INFO Consuming "Openshift Manifests" from target directory
INFO Consuming "Common Manifests" from target directory
INFO Creating infrastructure resources...
....
```

and all i'm saying is this is not a change we should be making now.

Reply:
With cf2e33350, the workflow you have above is working again.